### PR TITLE
fix: improve cached image startup and cache features

### LINKED ISSUE #334: Dev Container features are not cached
Extracting from https://github.com/coder/team-coconut/issues/39#issuecomment-2289825772

I leverage full image caching but have dev container features. https://github.com/coder/envbuilder/issues/210 seems to be one oe the culprits.

### PR DESCRIPTION
**Pull request by mafredri**
Feature files are cached in `/.envbuilder/features`, including their devcontainer JSON configuration. These JSON files can later be used to compute a fully fledged devcontainer file that combines project and features.

We also fix startup from an envbuilder built image so that we don't accidentally re-build the image.

Fixes #334

(I recommend reviewing with ignore whitespace on: https://github.com/coder/envbuilder/pull/353/files?w=1)

### CODE REVIEW

File: integration/integration_test.go [hunk 1]
```diff
@@ -1210,90 +1215,33 @@ RUN date --utc > /root/date.txt`, testImageAlpine),
...
-		require.Regexp(t, `ENVBUILDER_CACHED_IMAGE=(\S+)`, string(logBytes))
 
-		// When: we pull the image we just built
-		rc, err := cli.ImagePull(ctx, ref.String(), image.PullOptions{})
-		require.NoError(t, err)
-		t.Cleanup(func() { _ = rc.Close() })
-		_, err = io.ReadAll(rc)
-		require.NoError(t, err)
+		// Then: re-running envbuilder with GET_CACHED_IMAGE should succeed
+		cachedRef := getCachedImage(ctx, t, cli, opts...)
```
Comment (mafredri) [inline integration/integration_test.go:1241]: (Review: Removing a bunch of duplication, logic changes are not intentional)

File: integration/integration_test.go [hunk 2]
```diff
@@ -155,8 +157,17 @@ FROM a AS b`, testImageUbuntu),
...
+	// Check that envbuilder started command as user.
+	// Since envbuilder starts as root, probe for up to 10 seconds.
+	for i := 0; i < 10; i++ {
+		out := execContainer(t, ctr, "ps aux | awk '/^pickme * 1 / {print $1}' | sort -u")
+		got := strings.TrimSpace(out)
+		if got == "pickme" {
+			return
+		}
+		time.Sleep(time.Second)
+	}
```
Comment (johnstcn) [inline integration/integration_test.go:169]: 👍

File: envbuilder.go [hunk 3]
```diff
@@ -276,407 +276,421 @@ func Run(ctx context.Context, opts options.Options) error {
...
-		// we fallback to whatever the `DefaultImage` is.
-		var err error
-		buildParams, err = defaultBuildParams()
-		if err != nil {
-			return fmt.Errorf("no Dockerfile or devcontainer.json found: %w", err)
+	var (
+		username       string
+		skippedRebuild bool
+	)
+	if _, err := os.Stat(magicDir.Image()); errors.Is(err, fs.ErrNotExist) {
```
Comment (mtojek) [inline envbuilder.go:283]: nit: that's a huge condition :) but I'm afraid it matches the rest of the codebase.

File: envbuilder.go [hunk 4]
```diff
@@ -1122,11 +1136,11 @@ func RunCacheProbe(ctx context.Context, opts options.Options) (v1.Image, error)
...
 	}
 
-	// Also touch the magic file that signifies the image has been built!A
+	// Also write the magic file that signifies the image has been built!
 	magicImageDest := filepath.Join(magicTempDir, "image")
-	opts.Logger(log.LevelDebug, "touching magic image file at %q in build context %q", magicImageDest, magicTempDir)
-	if err := touchFile(opts.Filesystem, magicImageDest, 0o755); err != nil {
-		return nil, fmt.Errorf("touch magic image file at %q: %w", magicImageDest, err)
+	opts.Logger(log.LevelDebug, "writing magic image file at %q in build context %q", magicImageDest, magicTempDir)
+	if err := writeFile(opts.Filesystem, magicImageDest, 0o755, fmt.Sprintf("USER=%s\n", buildParams.User)); err != nil {
```
Comment (mtojek) [inline envbuilder.go:1142]: nit: you may drop a comment describing why `touch` was insufficient

**Review by mtojek — APPROVED** [review]
👍 

My main interest focused on how you fixed it, so I just skimmed through tests.

**Review by johnstcn — APPROVED** [review]
👍

File: envbuilder.go (hunk 3 above)
Comment (mafredri) [reply envbuilder.go:283]: Yes, agreed. Some refactoring in the future would be on-point. 😄

**Pull request merged**